public/tutorials/debugging-a-smart-contract.md

```diff
@@ -0,0 +1,621 @@
+
+   The output is as follows:
+
+   ```shell
+   Store.sol | 0x377bbcae5327695b32a1784e0e13bedc8e078c9c:
+
+   4:   uint myVariable;
+   5:
+   6:   function set(uint x) {
+        ^^^^^^^^^^^^^^^^^^^^^
+   ```
+
+   Notice that the program has moved to the next instruction, located on line 6. (The carats point to the exact part of the instruction taking place.)
+
+1. Press `Enter` again to step to the next instruction:
+
+   ```shell
+   Store.sol | 0x377bbcae5327695b32a1784e0e13bedc8e078c9c:
+
+   5:
+   6:   function set(uint x) {
+   7:     while(true){myVariable = x;}
+          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+   ```
+
+1. Press Enter three more times:
+
+   ```shell
+   Store.sol | 0x377bbcae5327695b32a1784e0e13bedc8e078c9c:
+
+   5:
+   6:   function set(uint x) {
+   7:     while(true){myVariable = x;}
+                      ^^^^^^^^^^
+
+   debug(develop:0xe4933407...)>
+
+   Store.sol | 0x377bbcae5327695b32a1784e0e13bedc8e078c9c:
+
+   5:
+   6:   function set(uint x) {
+   7:     while(true){myVariable = x;}
+                      ^^^^^^^^^^^^^^
+
+   debug(develop:0xe4933407...)>
+
+   Store.sol | 0x377bbcae5327695b32a1784e0e13bedc8e078c9c:
+
+   5:
+   6:   function set(uint x) {
+   7:     while(true){myVariable = x;}
+          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^   
+   ```
+
+   Notice that the bottom transaction is a repeat. In fact, pressing `Enter` over and over will repeat those last three transactions forever (or at least until the transaction runs out of gas). **This tells you where the problem is.**
+
+1. Type `q` to exit the debugger and return to the console.
+
+
+###  #2: An invalid error check
+
+Smart contracts can use statements like `assert()` to ensure that certain conditions are met. These can conflict with the state of the contract in ways that are irreconcilable.
+
+Here we will introduce such a condition, and then see how the debugger can find it.
+
+#### Introducing the error
+
+1. Open up the `Store.sol` file again.
+
+1. Replace the `set()` function with the following:
+
+   ```shell
+   function set(uint x) {
+     assert(x == 0);
+     myVariable = x;
+   }
+   ```
+
+   This is the same as the original version, but with an `assert()` function added, testing to make sure that `x == 0`. This will be fine until we set that value to something else, and then we will have a problem.
+
+#### Testing the contract
+
+Just as before, we'll close and reopen the console and redeploy the contract.
+
+1. Type `.exit` or press `Ctrl-C` twice in the Truffle Develop console to close it.
+
+1. Press `Ctrl-C` twice in the window showing the logs to close it.
+
+   <p class="alert alert-info">
+   <strong>Note</strong>: Remember that both consoles need to be closed first before starting a new session.
+   </p>
+
+1. In the first window, re-run `truffle develop`.
+
+1. In the second window, re-run `truffle develop --log`.
+
+1. In the first console, compile and migrate the contract:
+
+   ```shell
+   migrate
+   ```
```

Comment:
Since we discovered that `migrate --reset` works yesterday, we may be able to get away with telling people to use that instead of telling people to restart their `truffle develop` consoles. (Although your call – restarting everything is certainly more guaranteed to have less weirdness)

Reply:
Oh sweet, that simplifies so many things about this tutorial. It wasn't working when I tried it before, but I just tested it now and it worked just fine.